File: packages/typescript/src/adapters/xlsx.ts
// Optional .xlsx adapter (Seam 2 in the PRD): reads a single sheet — grid,
// origin, per-cell data type, and embedded chart descriptors — into the core's
// input contract (`Grid`, see ../types.ts).
//
// SheetJS (`xlsx`) is declared in package.json under `optionalDependencies`
// so the pure core can be installed without it. This module loads SheetJS
// lazily via createRequire so importing this file in a host that lacks it
// only fails when readSheet() is actually called.

import { createRequire } from "node:module";

import { a1 } from "../address.ts";
import type {
  CellMeta,
  ChartDescriptor,
  ChartType,
  DataType,
  Grid,
} from "../types.ts";

export type ReadSheetOptions = {
  /**
   * Which sheet to read. Pass a sheet name (string) or 0-indexed position
   * (number). Defaults to the workbook's first sheet.
   */
  sheet?: string | number;
};

/** A buffer-ish input the underlying tokenizer / xlsx parser accepts. */
export type ReadSheetInput = string | Buffer | Uint8Array | ArrayBuffer;

const req = createRequire(import.meta.url);

/* -------------------------------------------------------------------------- */
/* SheetJS — lazy load + minimal local typing                                 */
/* -------------------------------------------------------------------------- */

// Hand-typed subset of the SheetJS surface this adapter touches. We deliberately
// don't import xlsx's own types: the dep is optional and may not be installed.
type XlsxCell = {
  t: "s" | "n" | "b" | "d" | "e" | "z";
  v?: string | number | boolean | Date;
  w?: string;
  f?: string;
};
type XlsxSheet = Record<string, XlsxCell | unknown> & {
  "!ref"?: string;
};
type XlsxWorkbook = {
  SheetNames: string[];
  Sheets: Record<string, XlsxSheet>;
  keys?: string[];
  files?: Record<string, { content?: Buffer | Uint8Array }>;
};
type XlsxCellAddress = { r: number; c: number };
type XlsxRange = { s: XlsxCellAddress; e: XlsxCellAddress };
type XlsxModule = {
  read: (data: unknown, opts?: Record<string, unknown>) => XlsxWorkbook;
  readFile: (path: string, opts?: Record<string, unknown>) => XlsxWorkbook;
  utils: {
    decode_range: (s: string) => XlsxRange;
    decode_cell: (s: string) => XlsxCellAddress;
    encode_cell: (a: XlsxCellAddress) => string;
  };
};

function loadXlsx(): XlsxModule {
  try {
    return req("xlsx") as XlsxModule;
  } catch (cause) {
    throw new Error(
      "readSheet() requires the optional 'xlsx' (SheetJS) dependency. " +
        "Install it with `npm install xlsx`, or build the Grid yourself " +
        "and pass it directly to compress().",
      { cause },
    );
  }
}

/* -------------------------------------------------------------------------- */
/* Cell-type mapping                                                          */
/* -------------------------------------------------------------------------- */

const T_TO_DATA_TYPE: Record<XlsxCell["t"], DataType> = {
  s: "text",
  n: "number",
  b: "bool",
  d: "date",
  e: "error",
  z: "empty",
};

function inferDataType(cell: XlsxCell | undefined): DataType {
  if (!cell) return "empty";
  // Formula wins over type — a `=SUM()` cell is `formula` regardless of the
  // underlying evaluated type. The core only ever reads dataType (Phase 2
  // adds style flags) so collapsing here is safe.
  if (cell.f) return "formula";
  return T_TO_DATA_TYPE[cell.t];
}

function cellText(cell: XlsxCell | undefined): string {
  if (!cell) return "";
  if (cell.w !== undefined) return cell.w;
  const v = cell.v;
  if (v === undefined || v === null) return "";
  if (v instanceof Date) return v.toISOString();
  return String(v);
}

/* -------------------------------------------------------------------------- */
/* Chart extraction — walks workbook XML via the `bookFiles` payload          */
/* -------------------------------------------------------------------------- */

const CHART_BODY_TO_TYPE: Record<string, ChartType> = {
  barChart: "bar",
  bar3DChart: "bar",
  lineChart: "line",
  line3DChart: "line",
  pieChart: "pie",
  pie3DChart: "pie",
  doughnutChart: "pie",
  scatterChart: "scatter",
  bubbleChart: "scatter",
  areaChart: "area",
  area3DChart: "area",
};

type RelMap = Record<string, { type: string; target: string }>;

function parseRels(xml: string): RelMap {
  const out: RelMap = {};
  const re =
    /<Relationship\s+[^>]*?\bId="([^"]+)"[^>]*?\bType="([^"]+)"[^>]*?\bTarget="([^"]+)"[^>]*?\/?>/g;
  for (let m; (m = re.exec(xml)); ) {
    out[m[1]!] = { type: m[2]!, target: m[3]! };
  }
  return out;
}

/** Resolve a relationship Target against the part containing it. */
function resolveRelTarget(partPath: string, target: string): string {
  if (target.startsWith("/")) return target.replace(/^\/+/, "");
  const parentSegs = partPath.split("/");
  parentSegs.pop(); // drop the file
  const relSegs = target.split("/");
  while (relSegs[0] === "..") {
    relSegs.shift();
    parentSegs.pop();
  }
  return [...parentSegs, ...relSegs].join("/");
}

function decodeXmlText(s: string): string {
  return s
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&amp;/g, "&");
}

/** Concatenate every `<a:t>…</a:t>` text run inside the wrapper. */
function extractRichText(scope: string): string | undefined {
  const parts: string[] = [];
  const re = /<a:t[^>]*>([\s\S]*?)<\/a:t>/g;
  for (let m; (m = re.exec(scope)); ) parts.push(decodeXmlText(m[1]!));
  if (parts.length === 0) return undefined;
  return parts.join("");
}

/** First `<…:title>` block under `scope`; supports both `c:` and unprefixed. */
function extractTitle(scope: string): string | undefined {
  const m = /<(?:c:)?title>([\s\S]*?)<\/(?:c:)?title>/.exec(scope);
  return m ? extractRichText(m[1]!) : undefined;
}

function extractFieldText(xml: string, tag: string): string | undefined {
  const m = new RegExp(
    `<(?:c:)?${tag}>([\\s\\S]*?)<\\/(?:c:)?${tag}>`,
  ).exec(xml);
  if (!m) return undefined;
  return decodeXmlText(m[1]!.trim());
}

/** Normalise a chart cell-reference (`Sheet1!$B$2:$B$4` → `B2:B4`). */
function normalizeRange(ref: string): string {
  const bang = ref.lastIndexOf("!");
  const tail = bang >= 0 ? ref.slice(bang + 1) : ref;
  return tail.replace(/\$/g, "").replace(/^'|'$/g, "");
}

/** Everything `parseChartXml` can derive from a chart XML part. The two
 * fields it can't (`name`, `anchorRange`) live on the parent drawing/anchor. */
type ParsedChart = Omit<ChartDescriptor, "name" | "anchorRange">;

function parseChartXml(xml: string): ParsedChart {
  // Chart type: first known chart-body element under plotArea.
  let chartType: ChartType = "other";
  for (const [tag, t] of Object.entries(CHART_BODY_TO_TYPE)) {
    if (new RegExp(`<c:${tag}[ >]`).test(xml)) {
      chartType = t;
      break;
    }
  }

  // Title under <c:chart>.
  const chartScope = /<c:chart>([\s\S]*?)<\/c:chart>/.exec(xml)?.[1] ?? xml;
  const title = extractTitle(
    // Only the segment OUTSIDE plotArea — plotArea contains axis titles too.
    chartScope.replace(/<c:plotArea>[\s\S]*?<\/c:plotArea>/, ""),
  );

  // Axis titles: catAx/dateAx → x, valAx → y. First match of each wins.
  const xAxBlock =
    /<c:(?:catAx|dateAx)>([\s\S]*?)<\/c:(?:catAx|dateAx)>/.exec(xml)?.[1];
  const yAxBlock = /<c:valAx>([\s\S]*?)<\/c:valAx>/.exec(xml)?.[1];
  const axes: { x?: string; y?: string } = {};
  if (xAxBlock !== undefined) {
    const t = extractTitle(xAxBlock);
    if (t !== undefined) axes.x = t;
  }
  if (yAxBlock !== undefined) {
    const t = extractTitle(yAxBlock);
    if (t !== undefined) axes.y = t;
  }

  // Series: every <c:ser>…</c:ser>. Name from c:tx (literal c:v or strRef.c:f),
  // values from c:val/c:numRef/c:f (or c:cat for category-only charts).
  const series: string[] = [];
  const data: string[] = [];
  const serRe = /<c:ser>([\s\S]*?)<\/c:ser>/g;
  for (let m; (m = serRe.exec(xml)); ) {
    const ser = m[1]!;
    const txBlock = /<c:tx>([\s\S]*?)<\/c:tx>/.exec(ser)?.[1];
    if (txBlock !== undefined) {
      const literal = extractFieldText(txBlock, "v");
      if (literal !== undefined) {
        series.push(literal);
      } else {
        const cellRef = extractFieldText(txBlock, "f");
        if (cellRef !== undefined) series.push(normalizeRange(cellRef));
      }
    }
    const valBlock =
      /<c:val>([\s\S]*?)<\/c:val>/.exec(ser)?.[1] ??
      /<c:cat>([\s\S]*?)<\/c:cat>/.exec(ser)?.[1];
    if (valBlock !== undefined) {
      const cellRef = extractFieldText(valBlock, "f");
      if (cellRef !== undefined) data.push(normalizeRange(cellRef));
    }
  }

  const result: ParsedChart = { type: chartType };
  if (title !== undefined) result.title = title;
  if (axes.x !== undefined || axes.y !== undefined) result.axes = axes;
  if (series.length > 0) result.series = series;
  if (data.length > 0) result.dataRanges = data;
  return result;
}

function anchorRangeFromDrawing(scope: string): string | undefined {
  const m =
    /<xdr:from>[\s\S]*?<xdr:col>(\d+)<\/xdr:col>[\s\S]*?<xdr:row>(\d+)<\/xdr:row>[\s\S]*?<\/xdr:from>[\s\S]*?<xdr:to>[\s\S]*?<xdr:col>(\d+)<\/xdr:col>[\s\S]*?<xdr:row>(\d+)<\/xdr:row>[\s\S]*?<\/xdr:to>/.exec(
      scope,
    );
  if (!m) return undefined;
  const fc = Number(m[1]) + 1;
  const fr = Number(m[2]) + 1;
  const tc = Number(m[3]) + 1;
  const tr = Number(m[4]) + 1;
  return `${a1(fr, fc)}:${a1(tr, tc)}`;
}

function nameFromAnchor(scope: string): string {
  return /<xdr:cNvPr\s+[^>]*?\bname="([^"]*)"/.exec(scope)?.[1] ?? "";
}

function fileText(
  files: Record<string, { content?: Buffer | Uint8Array }>,
  path: string,
): string | undefined {
  const f = files[path];
  if (!f?.content) return undefined;
  return Buffer.from(f.content).toString("utf8");
}

function extractCharts(
  wb: XlsxWorkbook,
  sheetIndex: number,
): ChartDescriptor[] {
  const files = wb.files;
  if (!files) return [];
  const sheetPath = `xl/worksheets/sheet${sheetIndex + 1}.xml`;
  const sheetXml = fileText(files, sheetPath);
  if (sheetXml === undefined) return [];

  // Find the worksheet's drawing relId (only one drawing part per sheet in v0).
  const drawingRelId = /<drawing\s+[^>]*?r:id="([^"]+)"/.exec(sheetXml)?.[1];
  if (drawingRelId === undefined) return [];

  const sheetRelsPath = `xl/worksheets/_rels/sheet${sheetIndex + 1}.xml.rels`;
  const sheetRels = parseRels(fileText(files, sheetRelsPath) ?? "");
  const drawingRel = sheetRels[drawingRelId];
  if (!drawingRel) return [];

  const drawingPath = resolveRelTarget(sheetPath, drawingRel.target);
  const drawingXml = fileText(files, drawingPath);
  if (drawingXml === undefined) return [];

  const drawingRelsPath = drawingPath.replace(
    /([^/]+)$/,
    "_rels/$1.rels",
  );
  const drawingRels = parseRels(fileText(files, drawingRelsPath) ?? "");

  const charts: ChartDescriptor[] = [];
  const anchorRe =
    /<xdr:(?:twoCellAnchor|oneCellAnchor|absoluteAnchor)\b[\s\S]*?<\/xdr:(?:twoCellAnchor|oneCellAnchor|absoluteAnchor)>/g;
  for (let m; (m = anchorRe.exec(drawingXml)); ) {
    const anchor = m[0];
    const chartRelId = /<c:chart\s+[^>]*?r:id="([^"]+)"/.exec(anchor)?.[1];
    if (chartRelId === undefined) continue;
    const chartRel = drawingRels[chartRelId];
    if (!chartRel) continue;
    const chartPath = resolveRelTarget(drawingPath, chartRel.target);
    const chartXml = fileText(files, chartPath);
    if (chartXml === undefined) continue;

    const anchorRange = anchorRangeFromDrawing(anchor);
    if (anchorRange === undefined) continue;

    // parseChartXml omits any optional field with no value, so the spread
    // never leaks `undefined` keys into the descriptor.
    charts.push({
      ...parseChartXml(chartXml),
      name: nameFromAnchor(anchor),
      anchorRange,
    });
  }
  return charts;
}

/* -------------------------------------------------------------------------- */
/* Public entry point                                                         */
/* -------------------------------------------------------------------------- */

export function readSheet(
  input: ReadSheetInput,
  options: ReadSheetOptions = {},
): Grid {
  const xlsx = loadXlsx();
  const wb =
    typeof input === "string"
      ? xlsx.readFile(input, { cellDates: true, bookFiles: true })
      : xlsx.read(input, {
          type: ArrayBuffer.isView(input) ? "buffer" : "array",
          cellDates: true,
          bookFiles: true,
        });

  if (wb.SheetNames.length === 0) {
    throw new Error("readSheet(): workbook contains no sheets");
  }

  let sheetName: string;
  let sheetIndex: number;
  if (typeof options.sheet === "string") {
    const i = wb.SheetNames.indexOf(options.sheet);
    if (i < 0) {
      throw new Error(
        `readSheet(): sheet "${options.sheet}" not found in workbook ` +
          `(available: ${wb.SheetNames.join(", ")})`,
      );
    }
    sheetIndex = i;
    sheetName = options.sheet;
  } else {
    sheetIndex = options.sheet ?? 0;
    const found = wb.SheetNames[sheetIndex];
    if (found === undefined) {
      throw new Error(
        `readSheet(): sheet index ${sheetIndex} out of range ` +
          `(workbook has ${wb.SheetNames.length} sheet(s))`,
      );
    }
    sheetName = found;
  }

  const ws = wb.Sheets[sheetName]!;
  const ref = ws["!ref"];

  // Empty worksheet — emit an empty grid anchored at A1. cellMeta is omitted
  // (there are no cells to describe).
  let grid: Grid;
  if (typeof ref !== "string" || ref.length === 0) {
    grid = { rows: [], origin: { row: 1, col: 1 } };
  } else {
    const range = xlsx.utils.decode_range(ref);
    const rows: string[][] = [];
    const cellMeta: CellMeta[][] = [];
    for (let r = range.s.r; r <= range.e.r; r++) {
      const rowVals: string[] = [];
      const rowMeta: CellMeta[] = [];
      for (let c = range.s.c; c <= range.e.c; c++) {
        const addr = xlsx.utils.encode_cell({ r, c });
        const cell = ws[addr] as XlsxCell | undefined;
        rowVals.push(cellText(cell));
        rowMeta.push({ dataType: inferDataType(cell) });
      }
      rows.push(rowVals);
      cellMeta.push(rowMeta);
    }
    grid = {
      rows,
      origin: { row: range.s.r + 1, col: range.s.c + 1 },
      cellMeta,
    };
  }

  const charts = extractCharts(wb, sheetIndex);
  if (charts.length > 0) grid.charts = charts;
  return grid;
}

File: packages/python/sheet_compressor/adapters/xlsx.py
"""Optional .xlsx adapter (SPEC §8 / PRD Seam 2) — openpyxl-backed.

Reads a single sheet — grid, origin, per-cell data type, and embedded chart
descriptors — into the core's input contract (``Grid``, see SPEC §1). The pure
compression core stays dependency-free; this module is the only place openpyxl
is referenced.

openpyxl is declared under ``[project.optional-dependencies].xlsx`` in
``pyproject.toml``. ``read_sheet`` imports it lazily and raises a clear
``ImportError`` naming the missing dependency when it's absent (SPEC §8.2).

Usage::

    from sheet_compressor import compress
    from sheet_compressor.adapters.xlsx import read_sheet

    grid = read_sheet("workbook.xlsx")
    result = compress(grid)
"""

from __future__ import annotations

import io
import os
from datetime import date, datetime, time
from typing import Any, BinaryIO, List, Mapping, Optional, Union

from ..address import a1


ReadSheetInput = Union[str, bytes, bytearray, BinaryIO, os.PathLike]


_MISSING_DEP_MSG = (
    "read_sheet() requires the optional 'openpyxl' dependency. "
    "Install it with `pip install sheet-compressor[xlsx]` or "
    "`pip install openpyxl`, or build the Grid yourself and pass it to "
    "compress()."
)


def _load_openpyxl():
    try:
        import openpyxl  # type: ignore[import-not-found]
    except ImportError as cause:
        raise ImportError(_MISSING_DEP_MSG) from cause
    return openpyxl


# ---------------------------------------------------------------------------
# Cell value / data-type extraction
# ---------------------------------------------------------------------------


def _cell_text(value: Any) -> str:
    """Render an openpyxl cell value as a string for the Grid."""
    if value is None:
        return ""
    if isinstance(value, bool):
        # bool is a subclass of int — branch first so True/False don't get
        # stringified as "1"/"0".
        return "TRUE" if value else "FALSE"
    if isinstance(value, datetime):
        return value.isoformat()
    if isinstance(value, date):
        return value.isoformat()
    if isinstance(value, time):
        return value.isoformat()
    if isinstance(value, float):
        # Keep integer-valued floats compact: 3.0 → "3", matching the TS /
        # Go adapters (and SheetJS's default `w`).
        if value.is_integer():
            return str(int(value))
        return str(value)
    return str(value)


def _data_type(cell) -> str:
    """Map an openpyxl ``cell.data_type`` to the SPEC §1 vocabulary.

    openpyxl uses single-letter codes: ``s`` (text), ``n`` (number),
    ``b`` (bool), ``d`` (date), ``f`` (formula), ``e`` (error),
    ``inlineStr`` / ``str`` (text). A ``cell.data_type == "n"`` with
    ``cell.value is None`` is a gap — surface as ``empty``.
    """
    t = cell.data_type
    if t == "f":
        return "formula"
    if cell.value is None:
        return "empty"
    if t in ("s", "str", "inlineStr"):
        return "text"
    if t == "n":
        return "number"
    if t == "b":
        return "bool"
    if t == "d":
        return "date"
    if t == "e":
        return "error"
    return "text"


# ---------------------------------------------------------------------------
# Chart extraction
# ---------------------------------------------------------------------------


_CHART_CLASS_TO_TYPE = {
    "BarChart": "bar",
    "BarChart3D": "bar",
    "LineChart": "line",
    "LineChart3D": "line",
    "PieChart": "pie",
    "PieChart3D": "pie",
    "DoughnutChart": "pie",
    "ScatterChart": "scatter",
    "BubbleChart": "scatter",
    "AreaChart": "area",
    "AreaChart3D": "area",
}


def _normalize_range(ref: Optional[str]) -> Optional[str]:
    """Strip sheet qualifier and absolute markers: ``'Sheet1'!$B$2:$B$4`` → ``B2:B4``."""
    if not ref:
        return None
    s = ref
    bang = s.rfind("!")
    if bang >= 0:
        s = s[bang + 1 :]
    s = s.replace("$", "")
    s = s.strip("'")
    return s


def _rich_text(title_or_axis) -> Optional[str]:
    """Concatenate every ``<a:t>`` run inside a Title block.

    openpyxl exposes axis titles and chart titles as ``Title`` objects whose
    text lives at ``title.tx.rich.p[i].r[j].t``. Either ``tx`` or ``rich`` can
    be absent; treat any missing layer as "no title".
    """
    if title_or_axis is None:
        return None
    tx = getattr(title_or_axis, "tx", None)
    if tx is None:
        return None
    rich = getattr(tx, "rich", None)
    if rich is None:
        # Could be a strRef pointing at a cell — emit the resolved ref.
        ref = getattr(tx, "strRef", None)
        if ref is not None and getattr(ref, "f", None):
            return _normalize_range(ref.f)
        return None
    parts: List[str] = []
    for para in rich.p or []:
        for run in para.r or []:
            t = getattr(run, "t", None)
            if t:
                parts.append(t)
    if not parts:
        return None
    return "".join(parts)


def _anchor_range(anchor) -> Optional[str]:
    """A1 range for the chart's anchor.

    ``TwoCellAnchor`` carries both ``_from`` and ``to`` corners. ``OneCellAnchor``
    only has ``_from`` + an ``ext`` extent — collapse to a single-cell range,
    which still satisfies SPEC §1 ("A1 range") and tells the LLM where the
    chart sits. ``AbsoluteAnchor`` (pixel-positioned) has no cell anchor at
    all — return ``None`` to skip the descriptor's ``anchorRange``.
    """
    if anchor is None:
        return None
    frm = getattr(anchor, "_from", None)
    if frm is None:
        return None
    fr = int(frm.row) + 1
    fc = int(frm.col) + 1
    to = getattr(anchor, "to", None)
    if to is not None:
        tr = int(to.row) + 1
        tc = int(to.col) + 1
        return f"{a1(fr, fc)}:{a1(tr, tc)}"
    return f"{a1(fr, fc)}:{a1(fr, fc)}"


def _series_name(series) -> Optional[str]:
    tx = getattr(series, "tx", None)
    if tx is None:
        return None
    # SeriesLabel: literal v wins over strRef.
    v = getattr(tx, "v", None)
    if v is not None:
        return str(v)
    ref = getattr(tx, "strRef", None)
    if ref is not None and getattr(ref, "f", None):
        return _normalize_range(ref.f)
    return None


def _series_data_range(series) -> Optional[str]:
    """Pull the values reference off a Series.

    Cartesian charts put the y-values on ``series.val``, scatter / bubble
    charts may keep them on ``series.yVal``; category-only charts (e.g. pie)
    fall back to ``series.cat``. We try the same order as the TS / Go ports.
    """
    for attr in ("val", "yVal", "cat"):
        block = getattr(series, attr, None)
        if block is None:
            continue
        for ref_attr in ("numRef", "strRef"):
            ref = getattr(block, ref_attr, None)
            if ref is not None and getattr(ref, "f", None):
                return _normalize_range(ref.f)
    return None


def _chart_type(chart) -> str:
    name = type(chart).__name__
    return _CHART_CLASS_TO_TYPE.get(name, "other")


def _extract_charts(ws) -> List[dict]:
    # openpyxl doesn't surface the drawing's cNvPr name on its chart objects,
    # so descriptor.name is left empty here. SPEC §1 treats `name` as
    # informational, and the chart's `title` carries the human-meaningful label.
    out: List[dict] = []
    for chart in ws._charts or []:
        anchor_range = _anchor_range(getattr(chart, "anchor", None))
        if anchor_range is None:
            continue
        descriptor: dict = {
            "name": "",
            "type": _chart_type(chart),
            "anchorRange": anchor_range,
        }
        title = _rich_text(getattr(chart, "title", None))
        if title is not None:
            descriptor["title"] = title

        axes: dict = {}
        x_title = _rich_text(getattr(getattr(chart, "x_axis", None), "title", None))
        if x_title is not None:
            axes["x"] = x_title
        y_title = _rich_text(getattr(getattr(chart, "y_axis", None), "title", None))
        if y_title is not None:
            axes["y"] = y_title
        if axes:
            descriptor["axes"] = axes

        series: List[str] = []
        data: List[str] = []
        for ser in getattr(chart, "series", None) or []:
            name = _series_name(ser)
            if name is not None:
                series.append(name)
            rng = _series_data_range(ser)
            if rng is not None:
                data.append(rng)
        if series:
            descriptor["series"] = series
        if data:
            descriptor["dataRanges"] = data
        out.append(descriptor)
    return out


# ---------------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------------


def _open_workbook(openpyxl_mod, input_: ReadSheetInput):
    """Dispatch to ``load_workbook`` with the right input shape.

    Strings / PathLike are paths on disk. Raw bytes are wrapped in a
    ``BytesIO`` because openpyxl expects a file-like object. Anything already
    file-like is passed through.
    """
    if isinstance(input_, (bytes, bytearray)):
        return openpyxl_mod.load_workbook(io.BytesIO(bytes(input_)), data_only=False)
    if isinstance(input_, (str, os.PathLike)):
        return openpyxl_mod.load_workbook(os.fspath(input_), data_only=False)
    return openpyxl_mod.load_workbook(input_, data_only=False)


def read_sheet(
    input_: ReadSheetInput,
    options: Optional[Mapping[str, Any]] = None,
) -> dict:
    """Read a single sheet from an .xlsx file and return a ``Grid`` (SPEC §1).

    ``options["sheet"]`` selects the sheet by name (``str``) or 0-indexed
    position (``int``); omitting it picks the first sheet.

    The returned dict has ``rows``, ``origin``, and — when the source has any
    cells — ``cellMeta`` parallel to ``rows``. When the worksheet has embedded
    charts, ``charts`` is set to the list of ``ChartDescriptor`` dicts in
    document order.
    """
    openpyxl = _load_openpyxl()
    opts = options or {}

    wb = _open_workbook(openpyxl, input_)
    sheet_names = wb.sheetnames
    if not sheet_names:
        raise ValueError("read_sheet(): workbook contains no sheets")

    selector = opts.get("sheet")
    if isinstance(selector, str):
        if selector not in sheet_names:
            raise ValueError(
                f"read_sheet(): sheet {selector!r} not found in workbook "
                f"(available: {', '.join(sheet_names)})"
            )
        sheet_name = selector
    else:
        idx = 0 if selector is None else int(selector)
        if idx < 0 or idx >= len(sheet_names):
            raise ValueError(
                f"read_sheet(): sheet index {idx} out of range "
                f"(workbook has {len(sheet_names)} sheet(s))"
            )
        sheet_name = sheet_names[idx]

    ws = wb[sheet_name]
    grid = _build_grid(ws)
    charts = _extract_charts(ws)
    if charts:
        grid["charts"] = charts
    return grid


def _build_grid(ws) -> dict:
    """Materialise ``ws`` into the SPEC §1 Grid shape.

    The used range is openpyxl's ``ws.dimensions``. For a sheet with no
    written cells, openpyxl reports ``"A1:A1"`` AND ``min_row==max_row==1``
    AND ``cell.value is None``; we detect that and return an empty grid
    anchored at A1 with ``cellMeta`` omitted (SPEC §8.1 lets the adapter skip
    cellMeta when there is nothing to describe).
    """
    if ws.max_row is None or ws.max_column is None:
        return {"rows": [], "origin": {"row": 1, "col": 1}}

    min_row = ws.min_row
    min_col = ws.min_column
    max_row = ws.max_row
    max_col = ws.max_column

    # Detect a truly empty sheet — openpyxl still reports min_row=max_row=1
    # for a fresh workbook with no cells.
    if (
        min_row == 1
        and max_row == 1
        and min_col == 1
        and max_col == 1
        and ws.cell(row=1, column=1).value is None
        and ws.cell(row=1, column=1).data_type == "n"
    ):
        return {"rows": [], "origin": {"row": 1, "col": 1}}

    rows: List[List[str]] = []
    cell_meta: List[List[dict]] = []
    for r in range(min_row, max_row + 1):
        row_vals: List[str] = []
        row_meta: List[dict] = []
        for c in range(min_col, max_col + 1):
            cell = ws.cell(row=r, column=c)
            row_vals.append(_cell_text(cell.value))
            row_meta.append({"dataType": _data_type(cell)})
        rows.append(row_vals)
        cell_meta.append(row_meta)

    return {
        "rows": rows,
        "origin": {"row": min_row, "col": min_col},
        "cellMeta": cell_meta,
    }
